src/hugr/views/petgraph.rs

```diff
@@ -1,187 +1,250 @@
+where
+    T: HugrView,
+{
+    type EdgeType = petgraph::Directed;
+}
 
-            fn from_index(&self, ix: usize) -> Self::NodeId {
-                NodeIndex::new(ix).into()
-            }
-        }
+impl<'a, T> pv::NodeCount for PetgraphWrapper<'a, T>
+where
+    T: HugrView,
+{
+    fn node_count(&self) -> usize {
+        HugrView::node_count(self)
+    }
+}
 
-        impl<'a, Root, Base> pv::EdgeCount for $hugr<'a, Root, Base>
-        where
-            Root: NodeHandle,
-            Base: HugrInternals + HugrView,
-        {
-            fn edge_count(&self) -> usize {
-                HugrView::edge_count(self)
-            }
-        }
+impl<'a, T> pv::NodeIndexable for PetgraphWrapper<'a, T>
+where
+    T: HugrView,
+{
+    fn node_bound(&self) -> usize {
+        HugrView::node_count(self)
+    }
 
-        impl<'a, Root, Base> pv::Data for $hugr<'a, Root, Base>
-        where
-            Root: NodeHandle,
-            Base: HugrInternals + HugrView,
-        {
-            type NodeWeight = OpType;
-            type EdgeWeight = EdgeKind;
-        }
+    fn to_index(&self, ix: Self::NodeId) -> usize {
+        ix.index.into()
+    }
 
-        impl<'g, 'a, Root, Base> pv::IntoNodeIdentifiers for &'g $hugr<'a, Root, Base>
-        where
-            Root: NodeHandle,
-            Base: HugrInternals + HugrView,
-        {
-            type NodeIdentifiers = <$hugr<'a, Root, Base> as HugrView>::Nodes<'g>;
+    fn from_index(&self, ix: usize) -> Self::NodeId {
+        NodeIndex::new(ix).into()
+    }
+}
 
-            fn node_identifiers(self) -> Self::NodeIdentifiers {
-                self.nodes()
-            }
-        }
+impl<'a, T> pv::EdgeCount for PetgraphWrapper<'a, T>
+where
+    T: HugrView,
+{
+    fn edge_count(&self) -> usize {
+        HugrView::edge_count(self)
+    }
+}
 
-        impl<'g, 'a, Root, Base> pv::IntoNodeReferences for &'g $hugr<'a, Root, Base>
-        where
-            Root: NodeHandle,
-            'g: 'a,
-            Base: HugrInternals + HugrView,
-        {
-            type NodeRef = HugrNodeRef<'a>;
-            type NodeReferences =
-                MapWithCtx<<$hugr<'a, Root, Base> as HugrView>::Nodes<'a>, Self, HugrNodeRef<'a>>;
-
-            fn node_references(self) -> Self::NodeReferences {
-                self.nodes()
-                    .with_context(self)
-                    .map_with_context(|n, &hugr| HugrNodeRef::from_node(n, hugr))
-            }
-        }
+impl<'a, T> pv::Data for PetgraphWrapper<'a, T>
+where
+    T: HugrView,
+{
+    type NodeWeight = OpType;
+    type EdgeWeight = EdgeKind;
+}
 
-        impl<'g, 'a, Root, Base> pv::IntoNeighbors for &'g $hugr<'a, Root, Base>
-        where
-            Root: NodeHandle,
-            Base: HugrInternals + HugrView,
-        {
-            type Neighbors = <$hugr<'a, Root, Base> as HugrView>::Neighbours<'g>;
+impl<'g, 'a, T> pv::IntoNodeReferences for &'g PetgraphWrapper<'a, T>
```

Comment:
This is particularly good - the differences between the two impls (because of which I had to move this outside my macro in #501) seem to have gone